### PR TITLE
@types/node: enable strictnullchecks, fix resulting issues

### PR DESCRIPTION
**Pull request by akdor1154**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
  - module.parent (null -> undefined) - verified with repl on node 8.2.1. No mention of `null` in docs.
  - assert.throws Error type optional - https://nodejs.org/api/assert.html#assert_assert_throws_block_error_message ("if specified")
  - streams (_destroy callback)
    - https://nodejs.org/api/stream.html#stream_readable_destroy_err_callback
    - https://github.com/mcollina/node/blob/master/lib/internal/streams/destroy.js#L32 (on the Error parameter being optional_
  - streams (simple initializer functions being equivalent to _internal method overrides) -  https://nodejs.org/api/stream.html#stream_new_stream_readable_options
  - readline (data is ignored if key is specified) - https://nodejs.org/api/readline.html#readline_rl_write_data_key
  - inspector (properties are optional) - https://github.com/ChromeDevTools/devtools-protocol/blob/master/json/js_protocol.json (note className and exceptionDetails are both listed as optional)
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

Update @types/node to successfully work with strictNullChecks: true. Also var -> let in node-tests.ts. The first commit just fixes the "use before declaration" errors local to `node-tests`, the rest go on to fix the actual declaration problems detected as a result.

I am a bit confused that I was unable to find any existing discussion over this, sorry if I am unwittingly stomping all over existing drama.

 -- Jarrad

### CODE REVIEW

**Comment by akdor1154** [comment]
Open question - I've fixed up definitions for errbacks *in the tests* to be `(error: Error|null, result: Result)`. *In theory*, this is what Node errbacks do by convention. However, I don't think anybody would like to trawl through the node codebase and make sure that nobody ever ever sends an `undefined` in case of an error.

Hence, should the *definitions* of Node builtin modules be changed to pass `(error: Error | null | undefined)` in general? This question is probably too broad for this PR, but it is something we might have to think about if it's accepted. (As it might encourage user code like `if (err === null) { }`, which I wouldn't trust for reasons above)

File: types/node/index.d.ts [hunk 1]
```diff
@@ -5036,6 +5038,7 @@ declare module "stream" {
             wrap(oldStream: NodeJS.ReadableStream): Readable;
             push(chunk: any, encoding?: string): boolean;
             destroy(error?: Error): void;
+            _destroy(error: Error | null, cb: (e: Error | null) => any): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: actually `_destroy` gets the error passed to `destroy` so I think here we have and exception to the `Error | null` rule.
And I think the return type of `cb` should be `void`.

File: types/node/index.d.ts [hunk 2]
```diff
@@ -5101,16 +5104,17 @@ declare module "stream" {
...
         }
 
         export class Writable extends Stream implements NodeJS.WritableStream {
             writable: boolean;
             constructor(opts?: WritableOptions);
-            _write(chunk: any, encoding: string, callback: Function): void;
-            _destroy(err: Error, callback: Function): void;
+            _write(chunk: string | Buffer | any, encoding: string, callback: (e?: Error) => any): void;
+            _writev(chunks: Array<{ chunk: string | Buffer, encoding: string }>, callback: Function): void;
+            _destroy(error: Error | null, cb: (e: Error | null) => any): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: As in Readable.

File: types/node/index.d.ts [hunk 3]
```diff
@@ -123,7 +123,7 @@ interface NodeModule {
     id: string;
     filename: string;
     loaded: boolean;
-    parent: NodeModule | null;
+    parent: NodeModule | undefined;
```
Comment (Flarna) [inline types/node/index.d.ts]: I can remember that `parent` may be also `null` in some cases but I can't remember the exact use case.
Please adapt also the `class NodeJS.Module` accordingly.

File: types/node/index.d.ts (hunk 1 above)
Comment (akdor1154) [reply types/node/index.d.ts]: destroy contains the call `this._destroy(error || null)` - https://github.com/mcollina/node/blob/master/lib/internal/streams/destroy.js#L32 .

Comment (Flarna) [reply types/node/index.d.ts]: Ah yes, you are right.

**Review by Flarna — APPROVED** [review]

**Comment by andy-ms** [comment]
@akdor1154 Please fix the merge conflict.
As for `null` vs `null | undefined`, I would wait until you actually see `undefined` coming out of node, assuming that their documentation guarantees that `null` should be sent instead.

**Comment by akdor1154** [comment]
@andy-ms Documentation makes no such guarantee, *convention* states that null should be preferred.
Documentation varies module to module (and is sometimes entirely absent..), but for example `fs` [states](https://nodejs.org/api/fs.html#fs_file_system)
> The asynchronous form always takes a completion callback as its last argument. The arguments passed to the completion callback depend on the method, but the first argument is always reserved for an exception. If the operation was completed successfully, then the first argument will be null or undefined.

If this were my own project the way I would type this is add a
```ts
type NodeErrBack<T, E=Error> = (error?: null | Error, result: T) => void
```
and use it throughout `node/index.d.ts`. Would that approach be acceptable?

**Comment by andy-ms** [comment]
Sure, although I would write `error: Error | null | undefined` as I don't think an optional parameter can come first.

**Comment by akdor1154** [comment]
should I add this to this PR? (it'll be a pretty big change, touching any function in node that takes a callback...)

**Comment by andy-ms** [comment]
It can be in a separate PR.
The node types haven't updated to ts2.3 yet so they don't support default generics. We could increment the version but it would need a good reason.

**Comment by akdor1154** [comment]
k, shall rebase+squash then continue discussion in a new PR.

**Comment by akdor1154** [comment]
The failing lint in duplexer3 is a result of this pr. The reason is because the `cb` parameter in `stream._write` is now typed as `=> void` instead of `=> any`. The failing statement in duplexer3 is
```js
writable._write = (input, encoding, done) => {
  if (something) {
    return done();
  }
  ..
}
```
which is a common pattern. Does this suggest callbacks should be typed as `=> any` and not `=> void`?

**Comment by Flarna** [comment]
Besides the first argument in callbacks also the second one is optional - if the first one is not `null`.
Would be nice if this known dependency could be modeled somehow otherwise users have to use two checks or the `!` operator.
Maybe the default type should be `NodeJS.ErrnoException`, I have seen several complains that checking the optional `errno`/`code` properties is not allowed by typescript.
I fear stepping to typescript 2.3 could be cumbersome as it would imply that all the > 1000 dependent modules have to use also >=2.3 to my understanding.

**Comment by Flarna** [comment]
Regarding the issue with duplexer3: I would vote against use of `any` in callbacks just to make lint happy. Lint should help us not we should help lint.
I think we should try to avoid `any` where possible.

**Comment by andy-ms** [comment]
You don't have to change the type, you can just `done(); return;`.

**Comment by akdor1154** [comment]
@andy-ms my point is more that `return done()` is idiomatic, and I don't want to create a situation where valid and idiomatic code causes a lint error. I think I would tend towards Flarna's approach that this is more a lint configuration problem though.

**Comment by andy-ms** [comment]
It won't cause a lint error for people not using DT's lint configuration. So I agree with @Flarna to use the most accurate type possible, and users who don't want a lint error on `return done();` should not use that lint rule.

**Comment by RyanCavanaugh** [comment]
@akdor1154 Please fix the failures indicated in the Travis CI log.

**Comment by minestarks** [comment]
@akdor1154 could you fix the test failures?

**Comment by minestarks** [comment]
I'm closing this PR due to lack of activity. If you would still like to get in these changes, please open a new PR. Thank you!

**Pull request closed without merging**